back/boxtribute_server/business_logic/warehouse/box/mutations.py

```diff
@@ -156,78 +164,119 @@ def resolve_move_boxes_to_location(*_, update_input):
+            tag_errors.append(
+                {"id": tag_id, "error": ResourceDoesNotExist(name="Tag", id=tag_id)}
+            )
+            continue
+
+        if (error := authorize_tag(tag)) is not None:
+            tag_errors.append({"id": tag_id, "error": error})
+            continue
+
+        if for_unassigning:
+            # For unassigning it's valid to operate with deleted or type-mismatched tags
+            valid_tags.append(tag)
+            continue
+
+        if tag.deleted_on is not None:
+            tag_errors.append({"id": tag_id, "error": DeletedTag(name=tag.name)})
+            continue
+
+        if tag.type == TagType.Beneficiary:
+            tag_errors.append(
+                {"id": tag_id, "error": TagTypeMismatch(expected_type=TagType.Box)}
+            )
+            continue
+
+        valid_tags.append(tag)
+
+    tag_base_ids = {t.base_id for t in valid_tags}
+    # All requested tags must be registered in the same base
+    if len(tag_base_ids) > 1:
+        return [], tag_errors + [
+            {"id": tag_id, "error": TagBaseMismatch()} for tag_id in tag_ids
+        ]
+
+    if for_unassigning:
+        for tag in valid_tags:
+            if tag.deleted_on is not None:
+                # When a tag is deleted, it is also unassigned from any resource, hence
+                # in theory unassigning a deleted tag should not happen. However we have
+                # to deal with legacy data (there are 1.3k boxes assigned to deleted
+                # tags) and want to be notified about it.
+                emit_sentry_message(
+                    "User unassigned deleted tag from boxes",
+                    level="warning",
+                    extras={"tag_id": tag.id},
+                )
+            if tag.type == TagType.Beneficiary:
+                # When a tag type changes, it is also unassigned from any resource,
+                # hence in theory boxes assigned with a Beneficiary tag should not
+                # occur. However we have to deal with legacy data and want to be
+                # notified about it.
+                emit_sentry_message(
+                    "User unassigned Beneficiary-type tag from boxes",
+                    level="warning",
+                    extras={"tag_id": tag.id},
+                )
+
+    return [tag.id for tag in valid_tags], tag_errors
+
+
+@mutation.field("assignTagsToBoxes")
+def resolve_assign_tags_to_boxes(*_, update_input):
+    tag_ids = set(update_input["tag_ids"])
+    valid_tag_ids, tag_errors = _validate_tags(tag_ids)
+
+    if not valid_tag_ids:
+        return BoxesTagsOperationResult(
+            updated_boxes=[],
+            invalid_box_label_identifiers=[],
+            tag_error_info=tag_errors,
         )
-        .order_by(Box.id)
+
+    label_identifiers = set(update_input["label_identifiers"])
+    valid_boxes = execute_sql(
+        label_identifiers,
+        g.user.authorized_base_ids("stock:write"),
```

Comment:
possibly replace by base_id that all boxes and tags are in

Reply:
I can't replace it by the base that all boxes are in, because this _is_ the query to determine all valid boxes.

Solved it by passing the common base of all tags (guaranteed to be a single one due to prior validation), which means that any boxes in other bases are marked as invalid.